### DISCUSSION TITLE
Writing to encrypted path resulting in Permission denied

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [gramineproject/gramine#1958] [Q&A] [answered]
Hey Gramine team, 
Just a quick question please, such that I can to try and rule out a problem I'm currently facing,

Is it possible to write to an encrypted file using pwrite?

According to the pwrite man page:

> The pwrite() function shall be equivalent to write(), except that it writes into a given position without changing the file pointer. The first three arguments to pwrite() are the same as write() with the addition of a fourth argument offset for the desired position inside the file.

So I am wondering if this is dissallowed on Gramine, because if it is, then that explains it and I need to think on a alternative to patch it.

Also as additional notes, I also tested if the fd was valid and open to write to using:

`fcntl(fd, F_GETFD)`

But pwrite keeps returning EACCES

Any clues what might be wrong?
Can provide more details such as manifest if needed, just trying to rule out first the availability of pwrite on encrypted files.
Thank you!

**Comment by kailun-qin** [comment]
> Is it possible to write to an encrypted file using pwrite?
> So I am wondering if this is dissallowed on Gramine

I think this is supported.

> But pwrite keeps returning EACCES

`pf_write()` failure would generate `-EACCESS`: https://github.com/gramineproject/gramine/blob/6167772ea50f734757a8c3d93481ac69276af675/libos/src/fs/libos_fs_encrypted.c#L580-L584, would you pls share the verbose Gramine log and manifest? Thanks!

**Reply by dimakuv** [reply]
Yes, as @kailun-qin says: `pwrite()` system call is supported on Encrypted Files, there should be no specific issue.

**Reply by tiagorvmartins** [reply]
Okay thank you for confirming that kailun-qin and dimakuv, then that's not the reason for what I am experiencing.

Another question before diving deeper, would a rename of an encrypted file could be the culprit?

Because If I recall correctly, @dimakuv mentioning in a previous discussion, that file path and naming (metadata) is a requirement for encryption logic to succeed, and maybe that's the reason for what I saw previous in the logs:

```
warning: lib_AESGCMDecrypt failed: -6
ipf_init_existing_file: failed to decrypt metadata: -11
```

Explaining the later "Permission denied" down the flow of the application.

@kailun-qin I am about to have more logs on this, it takes some time to reach the point I need to have the relevant logs, once I get those I edit main post with more info

Thank you both!

**Reply by dimakuv** [reply]
> Another question before diving deeper, would a rename of an encrypted file could be the culprit?

How do you rename this encrypted file? The rename can only legitimately happen from **inside** the enclave (i.e., the application wants to rename). If the rename happens on the host (outside of the enclave), then yes, you'll observe the error.

**Reply by tiagorvmartins** [reply]
Sorry for the delay answering and thank you for the support, 

Been trying to solve it myself, redoing same steps several times and trying different approaches as well, but couldn't find solution or the culprit yet.

The rename happens inside the enclave, even though it's an application that creates several forks I believe that scenario wouldn't cause the issue correct? I even tried to use the path encrypted with the key _sgx_mrsigner instead of _sgx_mrenclave, but same issue.

Not using encrypted files is not an option as far as i know, because it relies on untrusted memory and allowed files disallow mmap according to what i have read before here: https://github.com/gramineproject/graphene/issues/433#issuecomment-864121459

I also understand the insecurity of using untrusted shared memory, I am quite aware of the implications that it raises.

I am adding now here the [manifest](https://pastebin.com/n4i0zTuu) and the full debug logs as attachment (long file) as @kailun-qin requested, I hope some of you two find what is going wrong here,

Logs are somewhat long, but I believe it's better this way so you have the entire context,
Thank you again for your support and time!

Note: just as side note, this is running on docker with gsc tooling (master branch)

[debug_logs.txt](https://github.com/user-attachments/files/16487197/debug_logs.txt)

**Reply by kailun-qin** [reply]
It looks like `/var/lib/postgresql/data/pg_wal/000000010000000000000001` was accessed concurrently by multiple processes but there is known limitation on this, pls see: https://github.com/gramineproject/gramine/blob/e14e18627bcafbfd1cc448fe3ff98ff685e4d6e4/Documentation/devel/encfiles.rst?plain=1#L500-L503.

**Reply by tiagorvmartins** [reply]
Thank you @kailun-qin , what would be the best approach to mitigate the issue?
Does using a file lock system using flock (fcntl primitive) on top of that file could suffice across those multiple processes using the same file?

**Reply by kailun-qin** [reply, marked as answer]
Yes, I think file locking for serialized access would work. Pls see the details w/ caveats below: https://github.com/gramineproject/gramine/blob/a9d08fde4abd374b762e23a8d8e3227265127913/Documentation/devel/features.md#L2225-L2258.
